core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala

```diff
@@ -424,6 +424,32 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with BeforeAndAfterE
     }
   }
 
+  test("deadlock between dropFromMemory and removeBlock") {
+    store = makeBlockManager(2000)
+    val a1 = new Array[Byte](400)
+    store.putSingle("a1", a1, StorageLevel.MEMORY_ONLY)
+    val t1 = new Thread {
+      override def run() = {
+        store.memoryManager.synchronized {
+          Thread.sleep(1000)
```

Comment:
I was able to reproduce the deadlock locally with this test almost every time, which is great. However, it would be good if you could rewrite this using explicit locks or semaphores or something so we don't have to introduce `Thread.sleep` here, which slows the test down and is more brittle.